Sources/PenguinGraphs/ApproximateKNN.swift

```diff
@@ -14,6 +14,50 @@
 
 import PenguinStructures
 
+extension IncidenceGraph where Self: VertexListGraph & MutableGraph {
+  /// Adds edges between `vertex` and its `k` nearest neighbors, as determined by `similarityBetween`.
+  ///
+  /// - Returns: A collection of the `k` nearest neighbors of `vertex` and their computed
+  ///   similarities.
+  /// - Complexity: O(|V| * O(similarityBetween)), where |V| is the number of vertices in `self`.
+  @discardableResult
+  public mutating func addKNearestNeighborEdges<Similarity: Comparable>(
+    vertex: VertexId,
+    k: Int,
+    similarityBetween: (VertexId, VertexId, inout Self) -> Similarity)
+  -> [(EdgeId, Similarity)] {
+    var workList = MaxPriorityQueue<VertexId, Similarity>()  // TODO: do truncation to `k`!
+    for u in vertices {
+      if u == vertex { continue }
+      let Similarity = similarityBetween(vertex, u, &self)
+      workList.push(u, at: Similarity)
+    }
+    var output = [(EdgeId, Similarity)]()
+    output.reserveCapacity(k)
+    var i = 0
+    while i < k {
+      i += 1
+      let vertexWithSimilarity = workList.pop()!
+      let edgeId = addEdge(from: vertex, to: vertexWithSimilarity.payload)
+      output.append((edgeId, vertexWithSimilarity.priority))
+    }
```

Comment:
Is there a reason this is a while loop and not a for loop? 

```
  for _ in 0..<k {
      let vertexWithSimilarity = workList.pop()!
      let edgeId = addEdge(from: vertex, to: vertexWithSimilarity.payload)
      output.append((edgeId, vertexWithSimilarity.priority))
    }
```

Reply:
Ah yup. Good call, that should be a for loop. Thanks!